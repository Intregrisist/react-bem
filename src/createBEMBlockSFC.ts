import * as PT from 'prop-types'
import * as React from 'react'

import { getDisplayName } from './helpers'
import { BEMBlockProps } from './types'
import resolveRenderedBlock from './resolveRenderedBlock'

export interface BEMBlockProviderContext {
	/**
	 * BEM block name
	 */
	bemBlock: string
}

/**
 * Wraps an SFC with BEM block functionality, providing the BEM block name
 * via context and converting block and modifiers attributes into className
 * attributes.
 * @param SFC The SFC to wrap with BEM block functionality.
 */
export default function createBEMBlockSFC<P ={}>(
	SFC: React.SFC<P>,
) {

	return class Wrapped<P2 = {}, S = {}>
	extends React.Component<any & P & P2 & BEMBlockProps, S> {

		static displayName = `BEMBlock(${getDisplayName(SFC as any)})`

		static childContextTypes = {
			bemBlock: PT.string.isRequired,
		}

		getChildContext(): BEMBlockProviderContext {
			return {
				bemBlock: this.props.bemBlock as string,
			}
		}

		render() {
			return resolveRenderedBlock(
				SFC(this.props, this.context),
				this.props,
			)
		}
	}
}
