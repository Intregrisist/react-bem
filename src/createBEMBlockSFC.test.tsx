import { render } from 'enzyme'
import * as PT from 'prop-types'
import * as React from 'react'

import createBEMBlockSFC from './createBEMBlockSFC'

describe('createBEMBlockSFC', () => {

	it('extends input class with a displayName of "BEMBlock(Foo)"', () => {
		const Foo: React.SFC = () => null
		const FooBlock = createBEMBlockSFC(Foo)
		expect(FooBlock.displayName).toEqual('BEMBlock(Foo)')
	})

	it('provides BEM block name via context', () => {
		const Foo: React.SFC = (props) => (
			<div>
				{props.children}
			</div>
		)
		const FooBlock = createBEMBlockSFC(Foo)
		const FooItem: React.SFC = (props, context) => {
			expect(context).toMatchSnapshot()
			return null
		}
		FooItem.contextTypes = {
			block: PT.string.isRequired,
		}
		render(
			<FooBlock block="context-block">
				<FooItem />
			</FooBlock>,
		)
	})

	it('merges modifiers prop with root node\'s modifiers', () => {
		const Foo: React.SFC = () => (
			<div modifiers="mod1" />
		)
		const FooBlock = createBEMBlockSFC(Foo)
		expect(render(
			<FooBlock block="foo" modifiers="mod2" />,
		)).toMatchSnapshot()
	})
})
