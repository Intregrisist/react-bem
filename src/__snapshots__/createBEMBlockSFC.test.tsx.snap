// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBEMBlockSFC merges modifiers prop with root node's modifiers 1`] = `
<div
  class="foo foo--mod1 foo--mod2"
/>
`;

exports[`createBEMBlockSFC provides BEM block name via context 1`] = `
Object {
  "block": "context-block",
}
`;
